packages/hydrogen/src/logger.ts

```diff
@@ -0,0 +1,103 @@
+import {yellow, red} from 'kolorist';
+
+type LoggerMethod = (
+  context: {request: Request},
+  ...args: Array<any>
+) => void | Promise<any>;
+
+interface LoggerOptions {
+  trace: LoggerMethod;
+  debug: LoggerMethod;
+  warn: LoggerMethod;
+  error: LoggerMethod;
+  fatal: LoggerMethod;
+}
+
+export class Logger {
+  #waitUntil: (p: Promise<unknown>) => void;
+  #request: Request;
+
+  constructor(
+    waitUntil: (p: Promise<unknown>) => void,
+    request: Request,
+    options: Partial<LoggerOptions> = {},
+  ) {
+    for (const [key, value] of Object.entries(options)) {
+      this.defaultLogger[key as keyof LoggerOptions] = value;
+    }
+    this.#waitUntil = waitUntil;
+    this.#request = request;
+  }
+
+  trace(...args: Array<unknown>) {
+    this.#doLog('trace', ...args);
+  }
+
+  debug(...args: Array<unknown>) {
+    this.#doLog('debug', ...args);
+  }
+
+  warn(...args: Array<unknown>) {
+    this.#doLog('warn', ...args);
+  }
+
+  error(...args: Array<unknown>) {
+    this.#doLog('error', ...args);
+  }
+
+  fatal(...args: Array<unknown>) {
+    this.#doLog('fatal', ...args);
+  }
+
+  #doLog(method: keyof LoggerOptions, ...args: Array<any>) {
+    try {
+      const maybePromise = this.defaultLogger[method](
+        {request: this.#request},
+        ...args,
+      );
+      if (maybePromise instanceof Promise) {
+        this.#waitUntil(
+          maybePromise.catch((e) => {
+            const message = e instanceof Error ? e.stack : e;
+            console.error(
+              `Promise error from the custom logging implementation for logger.${method} failed:\n${message}`,
+            );
+          }),
+        );
+      }
+    } catch (e) {
+      const message = e instanceof Error ? e.stack : e;
+      console.error(
+        `The custom logging implementation for logger.${method} failed:\n${message}`,
+      );
+    }
+  }
+
+  defaultLogger: LoggerOptions = {
+    trace: (context, ...args) => {
+      console.log(...args);
+    },
+
+    debug: (context, ...args) => {
+      console.log(...args);
+    },
+
+    warn: (context, ...args) => {
+      console.warn(yellow('WARN: '), ...args);
+    },
+
+    error: (context, error, ...extra) => {
+      const url = context ? ` ${context.request.url}` : '';
+
+      if (error instanceof Error) {
+        console.error(red(`Error processing route:${url}\n${error.stack}`));
+      } else {
+        console.error(red(`Error:${url} ${error}`));
+      }
```

Comment:
I copied this logic from H1, though not sure if we want to keep it 🤷🏼

Reply:
I wonder if this is where we implement the logging format that we discussed in #876?

Reply:
>I copied this logic from H1, though not sure if we want to keep it 🤷🏼 

Maybe we can remove it for now?

>I wonder if this is where we implement the logging format that we discussed in 876

I think that would be a different place since that should only be for Hydrogen logs. This logger here is supposed to be used manually by the user as well, right?